hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/ClientCacheProxyFactory.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.cache.impl;
+
+import com.hazelcast.cache.CacheNotExistsException;
+import com.hazelcast.cache.HazelcastCacheManager;
+import com.hazelcast.client.impl.HazelcastClientInstanceImpl;
+import com.hazelcast.client.spi.ClientProxy;
+import com.hazelcast.client.spi.ClientProxyFactory;
+import com.hazelcast.config.CacheConfig;
+
+public class ClientCacheProxyFactory implements ClientProxyFactory {
+
+    private final HazelcastClientInstanceImpl client;
+
+    public ClientCacheProxyFactory(HazelcastClientInstanceImpl client) {
+        this.client = client;
+    }
+
+    @Override
+    public ClientProxy create(String id) {
+        CacheConfig cacheConfig = findCacheConfig(id);
+        if (cacheConfig == null) {
+            throw new CacheNotExistsException("Cache " + id + " is already destroyed or not created yet");
+        }
+        return new ClientCacheProxy(cacheConfig);
+    }
+
+    private CacheConfig findCacheConfig(String cacheName) {
+        /**
+         * If there is specified cache config for thread-local usage, get and use it.
+         *
+         * Thread-local cache config is set by {@link com.hazelcast.client.cache.impl.HazelcastClientCacheManager}
+         * for passing cache config indirectly here
+         * (called by {@link com.hazelcast.client.spi.ProxyManager} while creating client side cache proxy)
+         * because of these reasons:
+         *      - To prevent unnecessary remote call to server for requesting cache config
+         *      - To use same cache config object create by
+         *        {@link com.hazelcast.client.cache.impl.HazelcastClientCacheManager}
+         */
+        CacheConfig cacheConfig = ClientCacheHelper.getThreadLocalCacheConfig();
+        if (cacheConfig != null) {
+            return cacheConfig;
+        }
+        // Otherwise, request it from server.
+        String simpleCacheName = cacheName.substring(HazelcastCacheManager.CACHE_MANAGER_PREFIX.length());
+        return ClientCacheHelper.getCacheConfig(client, cacheName, simpleCacheName);
```

Comment:
Might be. In fact same code block had been used by `HazelcastClientCacheManager` already so I keep it as it is and moved to this helper class.